sw/host/opentitanlib/src/transport/mod.rs

```diff
@@ -92,6 +96,10 @@ pub trait Transport {
     fn fpga_program(&self, _bitstream: &[u8]) -> Result<()> {
         unimplemented!();
     }
+    /// Replace the content of executable flash memory, and start executing.
+    fn bootstrap(&self, _image_filename: &std::path::PathBuf) -> Result<()> {
+        unimplemented!();
+    }
```

Comment:
I've been thinking about a more generic mechanism for hooks and this might be something that would qualify.

We start with a trait method like:

```
fn dispatch(&self, action: &dyn Any) -> Result<...>  // Don't know what goes here yet.  Probably Option<Box<erased_serde::Serializable>>>.
```

Then you declare a struct which can capture what action you want to perform and any configuration you might want to give the transport for that action:

```
struct EmulatorBootstrap {
  image: PathBuf,
  <other stuff>
}
```
And we use downcasts in the backend to figure out what the user wants and do it:

```
In the cw310 backend:
impl Transport for CW310 {
  fn dispatch(&self, ...) -> Result<> {
    unimplemented!();
  }
}

In the H1Dx or verilator emulation backends, something like:

impl Transport for ... {
  fn dispatch(&self, action: &dyn Any) -> Result<> {
    if let (params) = action.downcast_ref::<EmulatorBootstrap>() {
      self.emulator_bootstrap(params);
    } else if let(magic) = action.downcast_ref::<SomeOtherMagic>() {
      self.emulator_magic(magic);
    } else {
      Err(NoDispatch(action.type_id()))
    }
  }
}
```

Finally, the command implementations need to know that they're invoking some custom dispatch, but for _most_ of them, I think that would be their normal behavior (bootstrap might be the exception in this regard - it's generic SPI for most backends, but special for a couple of emulator backends).

Reply:
This could work, and be a way to avoid having many separate methods in the Transport trait.  I attempted converting this method, and fpga_program() to fit that pattern.  The only annoying thing is that in order to satisfy the Any trait, the structs can only have 'static references, meaning that the entire bitstream array will have to be copied.

That may not be a big problem though.  I can upload an alternative version soon.